kubernetes-model-generator/kubernetes-model-core/src/main/java/io/fabric8/kubernetes/api/model/HasMetadata.java

```diff
@@ -493,4 +496,26 @@ default void removeOwnerReference(HasMetadata owner) {
   default Optional<ObjectMeta> optionalMetadata() {
     return Optional.ofNullable(getMetadata());
   }
+
+  static <T extends HasMetadata & Namespaced> T emptyCopyOf(T original, Supplier<T> constructor) {
```

Comment:
Instead of emitCopyOf (copy could be missleading here) could this be named something like withNewMetadata? That would make it consistent with what it is doing and existing builder methods.

Reply:
Well, it is a copy. Not sure how that would be misleading?

Reply:
How about `newNamedInstance` and `newNamedAndNamespacedInstance`?

Reply:
It's only a copy or a new instance if the Supplier provides a new instance / copy. Even then if it's just the constructor, then it's a shallow copy.

If you want a full copy maybe specifying Editable would be the way to go.

Reply:
The goal is to get a shallow copy with only name/namespace set. Indeed, that will depend on the provided Supplier but I guess that also provides some flexibility as to what people want to initialize their instance with. Another "interesting" issue is that it seems we can't constraint the Supplier to return the exact object type as defined by the parameter type, which is rather annoying.

Reply:
In that case maybe simplify to just:

```
  static ObjectMeta newObjectMetaLike(HasMetadata original) {
    final ObjectMeta metadata = Objects.requireNonNull(original.getMetadata(), REQUIRES_NON_NULL_METADATA);
    ObjectMeta objectMeta = new ObjectMeta();
    objectMeta.setName(Objects.requireNonNull(metadata.getName(), REQUIRES_NON_NULL_NAME));
    if (Utils.isResourceNamespaced(original.getClass()) {
      objectMeta.setNamespace(Objects.requireNonNull(metadata.getNamespace(), REQUIRES_NON_NULL_NAMESPACE));
    }
    return objectMeta;
  }
```

And leave it up to the caller how they associate it with an instance.

Reply:
Makes sense but then I'd argue that this method should be put on `ObjectMeta` itself.

Reply:
We'd have to remove the ObjectMeta from automatic generation - we don't yet have the ability to mix-in additional methods.

Reply:
If it's not on `ObjectMeta`, I'd be concerned about discoverability. That said, we could also maybe have a `initNameAndNamespaceFrom` method that would set a new metadata on the passed object and retain said object?

```java
static HasMetadata initNameAndNamespaceFrom(HasMetadata original, HasMetata instanceToInit) {
    final ObjectMeta metadata = Objects.requireNonNull(original.getMetadata(), REQUIRES_NON_NULL_METADATA);
    ObjectMeta objectMeta = new ObjectMeta();
    objectMeta.setName(Objects.requireNonNull(metadata.getName(), REQUIRES_NON_NULL_NAME));
    if (original instanceof Namespaced) {
      objectMeta.setNamespace(Objects.requireNonNull(metadata.getNamespace(), REQUIRES_NON_NULL_NAMESPACE));
    }
    instanceToInit.setMetadata(objectMetadata);
    return instanceToInit;
  }
```

Reply:
So which version should I keep?

Reply:
@metacosm sorry I missed the previous message. I'm fine with what you have. initNameAndNamespaceFrom could also be done directly on HasMetadata:

```
default void initNameAndNamespaceFrom(HasMetadata original) {
    final ObjectMeta metadata = Objects.requireNonNull(original.getMetadata(), REQUIRES_NON_NULL_METADATA);
    ObjectMeta objectMeta = new ObjectMeta();
    objectMeta.setName(Objects.requireNonNull(metadata.getName(), REQUIRES_NON_NULL_NAME));
    if (original instanceof Namespaced) {
      objectMeta.setNamespace(Objects.requireNonNull(metadata.getNamespace(), REQUIRES_NON_NULL_NAMESPACE));
    }
    this.setMetadata(objectMeta);
  }
```

The only other question or thought is whether you are typically starting with builders instead, which would probably look like:

```
static ObjectMetaBuilder initNameAndNamespaceFrom(ObjectMetaBuilder builder, HasMetadata original) {
  ...
}
```

Reply:
Ah yes, good points!

Reply:
Should be good now, I think…